### PR TITLE
Spring Boot practice

### PR DESCRIPTION
**Pull request by aksoli666**
+added ShoppingCart and CartItem model

### CODE REVIEW

File: src/main/java/online/book/store/model/CartItem.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+@Entity
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @NotNull
+    @ManyToOne
```
Comment (kerrrusha) [inline src/main/java/online/book/store/model/CartItem.java:29]: join column?

File: src/main/java/online/book/store/model/CartItem.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+@ToString
+@Entity
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @NotNull
```
Comment (kerrrusha) [inline src/main/java/online/book/store/model/CartItem.java:28]: is redundant. such annotations are used for validation DTOs, not entities
```suggestion
```

File: src/main/java/online/book/store/model/CartItem.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    @ManyToOne
+    private ShoppingCart shoppingCart;
+    @NotNull
+    @ManyToOne
+    private Book book;
+    @NotNull
```
Comment (kerrrusha) [inline src/main/java/online/book/store/model/CartItem.java:34]:
```suggestion
    @ManyToOne
    private Book book;
    @Column(nullable = false)
```

File: src/main/java/online/book/store/model/CartItem.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @NotNull
+    @ManyToOne
+    private ShoppingCart shoppingCart;
+    @NotNull
+    @ManyToOne
```
Comment (kerrrusha) [inline src/main/java/online/book/store/model/CartItem.java:32]: join column?

File: src/main/java/online/book/store/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+@ToString
+@Entity
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @NotNull
```
Comment (kerrrusha) [inline src/main/java/online/book/store/model/ShoppingCart.java:31]:
```suggestion
```

File: src/main/java/online/book/store/repository/ShoppingCartRepository.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package online.book.store.repository;
+
+import online.book.store.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+
```
Comment (kerrrusha) [inline src/main/java/online/book/store/repository/ShoppingCartRepository.java:7]:
```suggestion
```

**Review by kerrrusha — CHANGES_REQUESTED** [review]

File: src/main/java/online/book/store/repository/ShoppingCartRepository.java (hunk 6 above)
Comment (aksoli666) [reply src/main/java/online/book/store/repository/ShoppingCartRepository.java:7]: Why do i need to remove this interfaces? i am using it

File: src/main/java/online/book/store/controller/CartController.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    @GetMapping
+    @Operation(summary = "Get a shopping cart", description = "Get a shopping cart by user id")
+    public ShoppingCartDto findByUserId(Long id) {
+        return cartService.findByUserId(id);
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PostMapping
+    @Operation(summary = "Add a book", description = "Add a book to shopping cart")
+    public ShoppingCartAddBookDto addBook(
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:36]:
```suggestion
    public ShoppingCartDto addBook(Authentication authentication,
                                   @RequestBody @Valid CartItemRequestDto cartItem)
```

File: src/main/java/online/book/store/controller/CartController.java [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            ShoppingCartAddBookDto shoppingCartDto,
+            ItemCartAddBookRequestDto itemCartDto) {
+        return cartService.addBook(shoppingCartDto, itemCartDto);
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:39]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return cartService.addBookToCart(user.getId(), cartItem);
```

File: src/main/java/online/book/store/controller/CartController.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @PutMapping
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:43]:
```suggestion
    @PutMapping("/cart-items/{cartItemId}")
```

File: src/main/java/online/book/store/controller/CartController.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    @Operation(summary = "Update a cart item", description = "Update a cart item by id")
+    public void updateCartItem(Long id, int quantity) {
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:45]:
```suggestion
    public void updateCartItem(Authentication authentication,
                                              @PathVariable @Positive Long cartItemId,
                                              @RequestBody @Valid UpdateCartItemDto item) {
```

This method should return updated shopping cart

File: src/main/java/online/book/store/controller/CartController.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        cartService.updateCartItem(id, quantity);
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:46]:
```suggestion
        User user = (User) authentication.getPrincipal();
        return cartService.updateBookQuantity(user.getId(), cartItemId, item);
```

File: src/main/java/online/book/store/controller/CartController.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @PreAuthorize("hasRole('ROLE_USER')")
+    @DeleteMapping
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:50]:
```suggestion
    @DeleteMapping("/cart-items/{cartItemId}")
```

File: src/main/java/online/book/store/controller/CartController.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    @Operation(summary = "Delete a book", description = "Delete a book")
+    public void deleteBook(Long id, CartItemDeleteBookDto deleteBookDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:52]: Did you check your endpoint? Can you provide id or dto object?

Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:52]:
```suggestion
    public void deleteBook(Authentication authentication,
                                            @PathVariable @Positive Long cartItemId) {
```

File: src/main/java/online/book/store/controller/CartController.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        cartService.deleteBook(id, deleteBookDto);
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/controller/CartController.java:53]:
```suggestion
        User user = (User) authentication.getPrincipal();
        cartService.deleteBookFromCart(user.getId(), cartItemId);
```

File: src/main/java/online/book/store/mapper/CartMapper.java [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+import online.book.store.dto.cart.request.ItemCartAddBookRequestDto;
+import online.book.store.model.CartItem;
+import online.book.store.model.ShoppingCart;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface CartMapper {
+    ShoppingCartDto toDto(ShoppingCart cart);
+
+    CartItemDto toDto(CartItem cartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/mapper/CartMapper.java:15]: It's not related to CartMapper

File: src/main/java/online/book/store/mapper/CartMapper.java [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+    ShoppingCart toModel(ShoppingCartDto cartDto);
+
+    CartItem toModel(CartItemDto cartItemDto);
+
+    CartItem toModel(ItemCartAddBookRequestDto requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/mapper/CartMapper.java:21]: the same as above

File: src/main/java/online/book/store/mapper/CartMapper.java [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+@Mapper(config = MapperConfig.class)
+public interface CartMapper {
+    ShoppingCartDto toDto(ShoppingCart cart);
+
+    CartItemDto toDto(CartItem cartItem);
+
+    ShoppingCart toModel(ShoppingCartDto cartDto);
+
+    CartItem toModel(CartItemDto cartItemDto);
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/mapper/CartMapper.java:19]: the same as above

File: src/main/java/online/book/store/model/CartItem.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+@ToString
+@Entity
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/model/CartItem.java:28]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "shopping_cart_id", nullable = false)
```

File: src/main/java/online/book/store/model/CartItem.java [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    @JoinColumn(name = "shopping_cart_id", referencedColumnName = "id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/model/CartItem.java:31]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "book_id", nullable = false)
```

File: src/main/java/online/book/store/model/ShoppingCart.java [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+@ToString
+@Entity
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = TRUE WHERE id = ?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/model/ShoppingCart.java:31]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/online/book/store/model/ShoppingCart.java [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+    @PrimaryKeyJoinColumn(name = "user_id", referencedColumnName = "id")
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/model/ShoppingCart.java:34]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/online/book/store/repository/BookRepository.java [hunk 22]
```diff
@@ -1,8 +1,14 @@
...
 import online.book.store.model.Book;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.data.jpa.repository.Query;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+    @Query(value = "SELECT * FROM books WHERE book_id IN ("
+            + "SELECT book_id FROM books_categories WHERE category_id = ?)",
+            nativeQuery = true)
+    List<Book> findBooksByCategoryId(Long categoryId);
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/repository/BookRepository.java:13]: I didn't find any logic related to Category. Don't need to create PR with several solutions

File: src/main/java/online/book/store/repository/ShoppingCartRepository.java (hunk 6 above)
Comment (Elena-Bruyako) [reply src/main/java/online/book/store/repository/ShoppingCartRepository.java:7]: You don't need to remove nothing here, just blank line

File: src/main/java/online/book/store/service/impl/CartServiceImpl.java [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+        return cartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public ShoppingCartAddBookDto addBook(
+            ShoppingCartAddBookDto shoppingCartDto,
+            ItemCartAddBookRequestDto itemCartDto) {
+        Set<CartItem> cartItems = shoppingCartDto.getCartItems();
+        cartItems.add(cartMapper.toModel(itemCartDto));
+        return shoppingCartDto;
```
Comment (Elena-Bruyako) [inline src/main/java/online/book/store/service/impl/CartServiceImpl.java:42]: Would you like to save the book into DB? What if this book is absent in DB? 
Please, implement the logic according to the lecture and video

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Need to watch video and lecture again and implement the right logic

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Pull request merged**